### ISSUE TITLE
Bogus DNSSEC error reported when mode is allow-downgrade

### ISSUE THREAD

**Issue by samueltardieu** [systemd/systemd#7421]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> systemd 235
> +PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN default-hierarchy=hybrid

### Used distribution

> ArchLinux x86_64

### In case of bug report: Expected behaviour you didn't see

`i.redd.it` should be resolved. It is not. Domains other than `redd.it` don't seem to be affected as far as my usage is concerned.

> % systemd-resolve i.redd.it          
> i.redd.it: resolve call failed: DNSSEC validation failed: failed-auxiliary

`journalctl` contains:

```
Nov 22 17:34:28 [hostname] systemd-resolved[875]: DNSSEC validation failed for question redd.it IN SOA: failed-auxiliary
Nov 22 17:34:28 [hostname] systemd-resolved[875]: DNSSEC validation failed for question i.redd.it IN A: failed-auxiliary
Nov 22 17:34:28 [hostname] systemd-resolved[875]: DNSSEC validation failed for question i.redd.it IN AAAA: failed-auxiliary
```

It can be properly resolved using `dig`:

```
% dig i.redd.it any
;; ANSWER SECTION:
i.redd.it.		286	IN	CNAME	reddit.map.fastly.net.

;; AUTHORITY SECTION:
redd.it.		3989	IN	NS	ns-673.awsdns-20.net.
redd.it.		3989	IN	NS	ns-344.awsdns-43.com.
redd.it.		3989	IN	NS	ns-1969.awsdns-54.co.uk.
redd.it.		3989	IN	NS	ns-1072.awsdns-06.org.

;; ADDITIONAL SECTION:
ns-344.awsdns-43.com.	51678	IN	A	205.251.193.88
ns-344.awsdns-43.com.	51678	IN	AAAA	2600:9000:5301:5800::1
ns-673.awsdns-20.net.	3480	IN	A	205.251.194.161
ns-1072.awsdns-06.org.	131281	IN	A	205.251.196.48
```

Note that domain `redd.it` does not use DNSSEC, and that even resolving the CNAME (to ensure that domain `fastly.net` don't participate to the problem here) won't work:

```
% systemd-resolve i.redd.it -t CNAME   
i.redd.it: resolve call failed: DNSSEC validation failed: failed-auxiliary
```

Also, `/etc/systemd/resolved.conf` contains:

```
[Resolve]
#DNS=
#FallbackDNS=8.8.8.8 8.8.4.4 2001:4860:4860::8888 2001:4860:4860::8844
#Domains=
#LLMNR=yes
DNSSEC=allow-downgrade
```

Changing it to `off` instead of `allow-downgrade` allows the system to resolve the name:

```
% systemd-resolve i.redd.it         
i.redd.it: 151.101.61.140
           (reddit.map.fastly.net)

-- Information acquired via protocol DNS in 12.3ms.
-- Data is authenticated: no
```

**Comment by ghost** [comment]
I am hitting the same problem (with the same host: i.redd.it). Here is my debug log.

```
Nov 23 09:46:38 fish systemd-resolved[1409]: Got message type=method_call sender=:1.30 destination=org.freedesktop.resolve1 object=/org/freedesktop/resolve1 interface=org.freedesktop.resolve1.Manager member=ResolveHostname cookie=2 reply_cookie=0 error=n/a
Nov 23 09:46:38 fish systemd-resolved[1409]: Looking up RR for i.redd.it IN A.
Nov 23 09:46:38 fish systemd-resolved[1409]: Looking up RR for i.redd.it IN AAAA.
Nov 23 09:46:38 fish systemd-resolved[1409]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=12 reply_cookie=0 error=n/a
Nov 23 09:46:38 fish systemd-resolved[1409]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.27 object=n/a interface=n/a member=n/a cookie=13 reply_cookie=12 error=n/a
Nov 23 09:46:38 fish systemd-resolved[1409]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=GetNameOwner cookie=13 reply_cookie=0 error=n/a
Nov 23 09:46:38 fish systemd-resolved[1409]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.27 object=n/a interface=n/a member=n/a cookie=14 reply_cookie=13 error=n/a
Nov 23 09:46:38 fish systemd-resolved[1409]: Cache miss for i.redd.it IN A
Nov 23 09:46:38 fish systemd-resolved[1409]: Transaction 48705 for <i.redd.it IN A> scope dns on */*.
Nov 23 09:46:38 fish systemd-resolved[1409]: Using feature level UDP+EDNS0+DO+LARGE for transaction 48705.
Nov 23 09:46:38 fish systemd-resolved[1409]: Using DNS server 108.61.10.10 for transaction 48705.
Nov 23 09:46:38 fish systemd-resolved[1409]: Sending query packet with id 48705.
Nov 23 09:46:38 fish systemd-resolved[1409]: Cache miss for i.redd.it IN AAAA
Nov 23 09:46:38 fish systemd-resolved[1409]: Transaction 52987 for <i.redd.it IN AAAA> scope dns on */*.
Nov 23 09:46:38 fish systemd-resolved[1409]: Using feature level UDP+EDNS0+DO+LARGE for transaction 52987.
Nov 23 09:46:38 fish systemd-resolved[1409]: Using DNS server 108.61.10.10 for transaction 52987.
Nov 23 09:46:38 fish systemd-resolved[1409]: Sending query packet with id 52987.
Nov 23 09:46:38 fish systemd-resolved[1409]: Processing incoming packet on transaction 48705. (rcode=SUCCESS)
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting parent SOA to validate transaction 48705 (i.redd.it, unsigned CNAME/DNAME/DS RRset).
Nov 23 09:46:38 fish systemd-resolved[1409]: Cache miss for redd.it IN SOA
Nov 23 09:46:38 fish systemd-resolved[1409]: Transaction 29739 for <redd.it IN SOA> scope dns on */*.
Nov 23 09:46:38 fish systemd-resolved[1409]: Using feature level UDP+EDNS0+DO+LARGE for transaction 29739.
Nov 23 09:46:38 fish systemd-resolved[1409]: Using DNS server 108.61.10.10 for transaction 29739.
Nov 23 09:46:38 fish systemd-resolved[1409]: Sending query packet with id 29739.
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting DS to validate transaction 48705 (fastly.net, unsigned SOA/NS RRset).
Nov 23 09:46:38 fish systemd-resolved[1409]: Cache miss for fastly.net IN DS
Nov 23 09:46:38 fish systemd-resolved[1409]: Transaction 20142 for <fastly.net IN DS> scope dns on */*.
Nov 23 09:46:38 fish systemd-resolved[1409]: Using feature level UDP+EDNS0+DO+LARGE for transaction 20142.
Nov 23 09:46:38 fish systemd-resolved[1409]: Using DNS server 108.61.10.10 for transaction 20142.
Nov 23 09:46:38 fish systemd-resolved[1409]: Sending query packet with id 20142.
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting DS to validate transaction 48705 (fastly.net, unsigned SOA/NS RRset).
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting DS to validate transaction 48705 (fastly.net, unsigned SOA/NS RRset).
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting DS to validate transaction 48705 (fastly.net, unsigned SOA/NS RRset).
Nov 23 09:46:38 fish systemd-resolved[1409]: Processing incoming packet on transaction 52987. (rcode=SUCCESS)
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting parent SOA to validate transaction 52987 (i.redd.it, unsigned CNAME/DNAME/DS RRset).
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting DS to validate transaction 52987 (fastly.net, unsigned SOA/NS RRset).
Nov 23 09:46:38 fish systemd-resolved[1409]: Processing incoming packet on transaction 29739. (rcode=SUCCESS)
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting DS to validate transaction 29739 (redd.it, unsigned SOA/NS RRset).
Nov 23 09:46:38 fish systemd-resolved[1409]: Cache miss for redd.it IN DS
Nov 23 09:46:38 fish systemd-resolved[1409]: Transaction 51052 for <redd.it IN DS> scope dns on */*.
Nov 23 09:46:38 fish systemd-resolved[1409]: Using feature level UDP+EDNS0+DO+LARGE for transaction 51052.
Nov 23 09:46:38 fish systemd-resolved[1409]: Using DNS server 108.61.10.10 for transaction 51052.
Nov 23 09:46:38 fish systemd-resolved[1409]: Sending query packet with id 51052.
Nov 23 09:46:38 fish systemd-resolved[1409]: Processing incoming packet on transaction 20142. (rcode=SUCCESS)
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting DNSKEY to validate transaction 20142 (A1RT98BS5QGC9NFI51S9HCI47ULJG6JH.net, RRSIG with key tag: 18737).
Nov 23 09:46:38 fish systemd-resolved[1409]: Cache miss for net IN DNSKEY
Nov 23 09:46:38 fish systemd-resolved[1409]: Transaction 56839 for <net IN DNSKEY> scope dns on */*.
Nov 23 09:46:38 fish systemd-resolved[1409]: Using feature level UDP+EDNS0+DO+LARGE for transaction 56839.
Nov 23 09:46:38 fish systemd-resolved[1409]: Using DNS server 108.61.10.10 for transaction 56839.
Nov 23 09:46:38 fish systemd-resolved[1409]: Sending query packet with id 56839.
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting DNSKEY to validate transaction 20142 (net, RRSIG with key tag: 18737).
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting DNSKEY to validate transaction 20142 (AP24DE1HQ24NHVG05VTUM96S2008U0H0.net, RRSIG with key tag: 18737).
Nov 23 09:46:38 fish systemd-resolved[1409]: Processing incoming packet on transaction 51052. (rcode=SUCCESS)
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting DNSKEY to validate transaction 51052 (5VIQIFPBOM04BTBL3TP9IV29Q63L76PU.it, RRSIG with key tag: 18395).
Nov 23 09:46:38 fish systemd-resolved[1409]: Positive cache hit for it IN DNSKEY
Nov 23 09:46:38 fish systemd-resolved[1409]: Transaction 50684 for <it IN DNSKEY> on scope dns on */* now complete with <success> from cache (authenticated).
Nov 23 09:46:38 fish systemd-resolved[1409]: Requesting DNSKEY to validate transaction 51052 (it, RRSIG with key tag: 18395).
Nov 23 09:46:38 fish systemd-resolved[1409]: Validating response from transaction 51052 (redd.it IN DS).
Nov 23 09:46:38 fish systemd-resolved[1409]: Looking at 5VIQIFPBOM04BTBL3TP9IV29Q63L76PU.it IN NSEC3 1 1 10 a9e3a37dda5c75f4 5VIQIFPBOM04BTBL3TP9IV29Q63L76PU ( NS SOA RRSIG DNSKEY NSEC3PARAM ): validated
Nov 23 09:46:38 fish systemd-resolved[1409]: Found verdict for lookup 5VIQIFPBOM04BTBL3TP9IV29Q63L76PU.it IN NSEC3: secure
Nov 23 09:46:38 fish systemd-resolved[1409]: Looking at it IN SOA dns.nic.it hostmaster.nic.it 2017112309 10800 900 604800 3600: validated
Nov 23 09:46:38 fish systemd-resolved[1409]: Found verdict for lookup it IN SOA: secure
Nov 23 09:46:38 fish systemd-resolved[1409]: Transaction 51052 for <redd.it IN DS> on scope dns on */* now complete with <EINVAL> from network (unsigned).
Nov 23 09:46:38 fish systemd-resolved[1409]: Auxiliary DNSSEC RR query failed with errno
Nov 23 09:46:38 fish systemd-resolved[1409]: DNSSEC validation failed for question redd.it IN SOA: failed-auxiliary
Nov 23 09:46:38 fish systemd-resolved[1409]: Transaction 29739 for <redd.it IN SOA> on scope dns on */* now complete with <dnssec-failed> from network (unsigned).
Nov 23 09:46:38 fish systemd-resolved[1409]: Auxiliary DNSSEC RR query failed validation: failed-auxiliary
Nov 23 09:46:38 fish systemd-resolved[1409]: DNSSEC validation failed for question i.redd.it IN A: failed-auxiliary
Nov 23 09:46:38 fish systemd-resolved[1409]: Transaction 48705 for <i.redd.it IN A> on scope dns on */* now complete with <dnssec-failed> from network (unsigned).
Nov 23 09:46:38 fish systemd-resolved[1409]: Auxiliary DNSSEC RR query failed validation: failed-auxiliary
Nov 23 09:46:38 fish systemd-resolved[1409]: DNSSEC validation failed for question i.redd.it IN AAAA: failed-auxiliary
Nov 23 09:46:38 fish systemd-resolved[1409]: Transaction 52987 for <i.redd.it IN AAAA> on scope dns on */* now complete with <dnssec-failed> from network (unsigned).
Nov 23 09:46:38 fish systemd-resolved[1409]: Freeing transaction 48705.
Nov 23 09:46:38 fish systemd-resolved[1409]: Sent message type=error sender=n/a destination=:1.30 object=n/a interface=n/a member=n/a cookie=14 reply_cookie=2 error=DNSSEC validation failed: failed-auxiliary
Nov 23 09:46:38 fish systemd-resolved[1409]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=RemoveMatch cookie=15 reply_cookie=0 error=n/a
Nov 23 09:46:38 fish systemd-resolved[1409]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.27 object=n/a interface=n/a member=n/a cookie=16 reply_cookie=15 error=n/a
Nov 23 09:46:38 fish systemd-resolved[1409]: Freeing transaction 52987.
Nov 23 09:46:38 fish systemd-resolved[1409]: Freeing transaction 20142.
Nov 23 09:46:38 fish systemd-resolved[1409]: Freeing transaction 56839.
Nov 23 09:46:38 fish systemd-resolved[1409]: Freeing transaction 29739.
Nov 23 09:46:38 fish systemd-resolved[1409]: Freeing transaction 51052.
Nov 23 09:46:38 fish systemd-resolved[1409]: Freeing transaction 50684.
Nov 23 09:46:38 fish systemd-resolved[1409]: Got message type=signal sender=org.freedesktop.DBus destination=n/a object=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameOwnerChanged cookie=15 reply_cookie=0 error=n/a
```

This is, what is currently being reported by [Verisign's DNSSEC Analyzer](https://dnssec-debugger.verisignlabs.com/i.redd.it).

![dnssec](https://user-images.githubusercontent.com/33265431/33167237-324a3d7c-d03d-11e7-83cd-1d74088880a8.png)

**Comment by samueltardieu** [comment]
The zone is `redd.it`, not `i.redd.it`. But note that the zone is not using DNSSEC, which is why the DNSSEC analyzer (which checks that the zone is properly configured with DNSSEC) reports an error.

The unsuccesful `DS` lookup in the `it` zone for `redd.it` is expected. However, I don't understand why it shows up as a failure in the log (`EINVAL`) instead of a zero-entry response.

```
Nov 23 09:46:38 fish systemd-resolved[1409]: Transaction 51052 for <redd.it IN DS> on scope dns on */* now complete with <EINVAL> from network (unsigned).
```

**Comment by samueltardieu** [comment]
It looks like all `.it` domains (CTLD for Italy) are affected:

```
% systemd-resolve turismoroma.it -t SOA
turismoroma.it: resolve call failed: DNSSEC validation failed: failed-auxiliary

% dig turismoroma.it SOA +short
ns1.register.it. hostmaster.register.it. 2017060603 10800 3600 604800 86400
```

**Comment by samueltardieu** [comment]
Here is the `delv` trace for `redd.it` `SOA` query. Note how it does not seem to consider any response as invalid (`EINVAL` in `systemd-resolved` log).

```
% delv redd.it -t SOA +vtrace
;; fetch: redd.it/SOA
;; validating redd.it/SOA: starting
;; validating redd.it/SOA: attempting insecurity proof
;; validating redd.it/SOA: checking existence of DS at 'it'
;; fetch: it/DS
;; validating it/DS: starting
;; validating it/DS: attempting positive response validation
;; fetch: ./DNSKEY
;; validating ./DNSKEY: starting
;; validating ./DNSKEY: attempting positive response validation
;; validating ./DNSKEY: verify rdataset (keyid=19036): success
;; validating ./DNSKEY: signed by trusted key; marking as secure
;; validating it/DS: in fetch_callback_validator
;; validating it/DS: keyset with trust secure
;; validating it/DS: resuming validate
;; validating it/DS: verify rdataset (keyid=46809): success
;; validating it/DS: marking as secure, noqname proof not needed
;; validating redd.it/SOA: in dsfetched2: success
;; validating redd.it/SOA: resuming proveunsecure
;; validating redd.it/SOA: checking existence of DS at 'redd.it'
;; fetch: redd.it/DS
;; validating redd.it/DS: starting
;; validating redd.it/DS: attempting negative response validation
;;   validating it/SOA: starting
;;   validating it/SOA: attempting positive response validation
;; fetch: it/DNSKEY
;; validating it/DNSKEY: starting
;; validating it/DNSKEY: attempting positive response validation
;; validating it/DNSKEY: verify rdataset (keyid=41901): success
;; validating it/DNSKEY: marking as secure (DS)
;;   validating it/SOA: in fetch_callback_validator
;;   validating it/SOA: keyset with trust secure
;;   validating it/SOA: resuming validate
;;   validating it/SOA: verify rdataset (keyid=18395): success
;;   validating it/SOA: marking as secure, noqname proof not needed
;; validating redd.it/DS: in authvalidated
;; validating redd.it/DS: resuming nsecvalidate
;;   validating 0IRC457CC3M11CT5SQ075F6E35SSAAG0.it/NSEC3: starting
;;   validating 0IRC457CC3M11CT5SQ075F6E35SSAAG0.it/NSEC3: attempting positive response validation
;;   validating 0IRC457CC3M11CT5SQ075F6E35SSAAG0.it/NSEC3: keyset with trust secure
;;   validating 0IRC457CC3M11CT5SQ075F6E35SSAAG0.it/NSEC3: verify rdataset (keyid=18395): success
;;   validating 0IRC457CC3M11CT5SQ075F6E35SSAAG0.it/NSEC3: marking as secure, noqname proof not needed
;; validating redd.it/DS: in authvalidated
;; validating redd.it/DS: resuming nsecvalidate
;; validating redd.it/DS: looking for relevant NSEC3
;; validating redd.it/DS: looking for relevant NSEC3
;; validating redd.it/DS: NSEC3 proves name does not exist: 'redd.it'
;; validating redd.it/DS: NSEC3 indicates optout
;; validating redd.it/DS: NSEC3 indicates potential closest encloser: 'it'
;; validating redd.it/DS: NSEC3 at super-domain it
;; validating redd.it/DS: in checkwildcard: *.it
;; validating redd.it/DS: looking for relevant NSEC3
;; validating redd.it/DS: NSEC3 proves name does not exist: '*.it'
;; validating redd.it/DS: NSEC3 at super-domain it
;; validating redd.it/DS: in checkwildcard: *.it
;; validating redd.it/DS: nonexistence proof(s) found
;; validating redd.it/SOA: in dsfetched2: ncache nxrrset
;; validating redd.it/SOA: plain DNSSEC returns unsecure (redd.it): looking for DLV
;; validating redd.it/SOA: looking for DLV redd.it.dlv.isc.org
;; fetch: redd.it.dlv.isc.org/DLV
;; validating redd.it/SOA: DLV lookup: wait
;; validating redd.it.dlv.isc.org/DLV: starting
;; validating redd.it.dlv.isc.org/DLV: attempting negative response validation
;;   validating dlv.isc.org/SOA: starting
;;   validating dlv.isc.org/SOA: attempting positive response validation
;; fetch: dlv.isc.org/DNSKEY
;; validating dlv.isc.org/DNSKEY: starting
;; validating dlv.isc.org/DNSKEY: attempting positive response validation
;; validating dlv.isc.org/DNSKEY: verify rdataset (keyid=19297): success
;; validating dlv.isc.org/DNSKEY: signed by trusted key; marking as secure
;;   validating dlv.isc.org/SOA: in fetch_callback_validator
;;   validating dlv.isc.org/SOA: keyset with trust secure
;;   validating dlv.isc.org/SOA: resuming validate
;;   validating dlv.isc.org/SOA: verify rdataset (keyid=64263): success
;;   validating dlv.isc.org/SOA: marking as secure, noqname proof not needed
;; validating redd.it.dlv.isc.org/DLV: in authvalidated
;; validating redd.it.dlv.isc.org/DLV: resuming nsecvalidate
;;   validating dlv.isc.org/NSEC: starting
;;   validating dlv.isc.org/NSEC: attempting positive response validation
;;   validating dlv.isc.org/NSEC: keyset with trust secure
;;   validating dlv.isc.org/NSEC: verify rdataset (keyid=64263): success
;;   validating dlv.isc.org/NSEC: marking as secure, noqname proof not needed
;; validating redd.it.dlv.isc.org/DLV: in authvalidated
;; validating redd.it.dlv.isc.org/DLV: looking for relevant NSEC
;; validating redd.it.dlv.isc.org/DLV: nsec range ok
;; validating redd.it.dlv.isc.org/DLV: resuming nsecvalidate
;; validating redd.it.dlv.isc.org/DLV: in checkwildcard: *.dlv.isc.org
;; validating redd.it.dlv.isc.org/DLV: looking for relevant NSEC
;; validating redd.it.dlv.isc.org/DLV: nsec range ok
;; validating redd.it.dlv.isc.org/DLV: nonexistence proof(s) found
;; validating redd.it/SOA: in dlvfetched: ncache nxdomain
;; validating redd.it/SOA: DLV not found
;; validating redd.it/SOA: marking as answer (dlvfetched (3))
; unsigned answer
redd.it.		900	IN	SOA	ns-344.awsdns-43.com. awsdns-hostmaster.amazon.com. 1 7200 900 1209600 86400
emu-rs/rustual-boy

```

**Comment by poettering** [comment]
Fix waiting in #7590
